optimization: Faster simulations by reducing number of rx_search operations

paulquevedojdrf opened the pull request:
Faster simulations by reducing number of rx_search operations

# Overview 

This PR attempts to speed up the time to simulate by reducing the number of function queue operations performed when devices are in the `Rx_Search` state. Currently they move in 1us steps, but if nothing is expected to happen for several microseconds, these are just wasted cycles resulting in longer simulation times.

Instead when a device executes an `Rx_Search`, it is put into a list of devices that are actively searching for a transmission. Once a transmission occurs, all actively searching devices are then moved into the `Rx_Sync` state.
This gets rid of the need to constantly re-queue the `Rx_Search` operation every microsecond. Instead its only queued once the scan time ends or for an abort check. 

# Testing

I  ran a simple simulation with 30 devices all randomly sending and receiving messages. The random seeds used are instance unique but static across runs (i.e. the device number). 

Compared the Tx and Rx dumps from the phy before and after this change and confirmed they are identical.

aescolar commented on src/p2G4_func_queue.c:
```diff
@@ -36,39 +45,57 @@ void fq_register_func(f_index_t type, queable_f fptr) {
  * Find the next function which should be executed,
```
This first line would not be applicable to the function below anymore

aescolar commented on src/p2G4_func_queue.c:
```diff
@@ -6,26 +6,35 @@
  * Each interface can have 1 function pending
  */
 static fq_element_t *f_queue = NULL;
+static uint32_t *device_list = NULL;
```
Would you be so kind as to comment shortly in the code these new variables?
(Also add a comment to each new functions)

aescolar commented on src/p2G4_func_queue.c:
```diff
@@ -36,39 +45,57 @@ void fq_register_func(f_index_t type, queable_f fptr) {
 void fq_add(bs_time_t time, f_index_t index, uint32_t dev_nbr) {
   fq_element_t *el = &f_queue[dev_nbr];
+  if ( el->time == time ) {
+    /* We are adding a new operation that must be run at the current time
```
>that must be run at the current time

You mean at the same time as this element that is just being replaced was meant to run(?)

paulquevedojdrf replied:
yes, Rx_Sync is added by Rx_Search using the same timestamp. wasn't sure if that was a bug or intentional but this effectively keeps the behavior the same.

aescolar replied:
>wasn't sure if that was a bug or intentional

To the best of my recollection, and note that this was 6 years ago :), intentional. The first bit of the packet is in air in that µs so we start checking if we receive it.

aescolar commented on src/p2G4_func_queue.c:
```diff
@@ -6,26 +6,35 @@
+static uint32_t *device_list = NULL;
 
-static uint32_t next_d = 0;
+static bs_time_t current_time = 0;
```
Should this be only in fq_step scope?

paulquevedojdrf replied:
yes but it becomes inaccessible from a unit-test perspective by doing so (i.e. reset to 0 on teardown). If its still needed after the proposed improvements I will move it

aescolar commented on src/p2G4_main.c:
```diff
@@ -183,7 +183,8 @@ static void f_rx_search(uint d) {
   }
 
-  fq_add(current_time + 1, Rx_Search, d);
+  bs_time_t expiry = BS_MIN(rx_status->scan_end + 1, rx_status->rx_s.abort.recheck_time);
```
A fundamental question: Would it not be easier conceptually, and hopefully faster, to get here the time of the next event in the queue and add to the queue this next RxSearch at that time? (without this new _pend/flush mechanism?) 
(Say, by modifying fq_call_next() so it sets the time of the element it is calling to TIME_NEVER, calling fq_find_next() here, and setting `expiry = MIM(expiry, fq_get_next_time());` )

paulquevedojdrf replied:
the time of the next event in the queue is not known until all events at the current time point have executed. There may be other devices that run after this one that will add a new event to the queue with an earlier timestamp than anything currently in the queue.

aescolar replied:
Yep.. what I proposed here would require rx_search to be the last type of event in the queue, and probably then adding a recheck @+1 if there is other elements to trigger now in the queue.. which for mesh is a certainty.. so it would give nothing for mesh...

aescolar reviewed (changes requested):
Thanks @paulquevedojdrf 
See the more fundamental question in the last comment.
Anyhow, I'd expect you would get a better*1 performance improvement by doing another optimization:
* Change the rx_search state to not try to find a matching Tx, but instead advance only to the scan end or abort recheck. Add a new field in `rx_status_t` (`searching`) which is set when entering rx_search, and removed when moving to another state. (Set rx_search as a higher event "priority" than tx start) And have the matching tx search be instead a matching rx search called from `txl_activate()`. When a match happens, the matched Rx is moved to the `rx_sync` state.

*1 To get a feeling of the improvement you could check in a profile run how many times txl_activate() is called today, vs how many times txl_find_fitting_tx() is called. The difference would be how many calls to f_rx_search (and txl_find_fitting_tx ) we would save.

See https://raw.githubusercontent.com/BabbleSim/ext_2G4_phy_v1/master/docs/Device_interface.svg

paulquevedojdrf commented:
> Thanks @paulquevedojdrf
> See the more fundamental question in the last comment.
> Anyhow, I'd expect you would get a better*1 performance improvement by doing another optimization:
> 
> * Change the rx_search state to not try to find a matching Tx, but instead advance only to the scan end or abort recheck. Add a new field in `rx_status_t` (`searching`) which is set when entering rx_search, and removed when moving to another state. (Set rx_search as a higher event "priority" than tx start) And have the matching tx search be instead a matching rx search called from `txl_activate()`. When a match happens, the matched Rx is moved to the `rx_sync` state.
> 
> *1 To get a feeling of the improvement you could check in a profile run how many times txl_activate() is called today, vs how many times txl_find_fitting_tx() is called. The difference would be how many calls to f_rx_search (and txl_find_fitting_tx ) we would save.

Yes I observed txl_find_fitting_tx() to be the next bottleneck in the simulation. I agree this is a better approach. I will try it out and profile it against the current implementation to see the gains. Will take me a couple days to find some time to do this though. 
Thanks for the review!

paulquevedojdrf committed:
optimization: Faster simulations by reducing number of rx_search operations

This commit changes the phy to track the devices currently
searching for a suitable transmission. Each time a transmission is
executed the devices in the search list attempt to synchronize. This
reduces the number of redundant function queue operations that occur as
a result of executing an rx search operation at every time step.

Signed-off-by: Paul Quevedo <paulquevedo@jdrfelectromag.com>

paulquevedojdrf commented:
Callgrinds showing same simulation profiled before and after this patch. Simulation results are identical. 

[callgrind.out.1356601-fast.txt](https://github.com/BabbleSim/ext_2G4_phy_v1/files/6827480/callgrind.out.1356601-fast.txt)
[callgrind.out.1356885-slow.txt](https://github.com/BabbleSim/ext_2G4_phy_v1/files/6827481/callgrind.out.1356885-slow.txt)

aescolar commented on src/p2G4_func_queue.c:
```diff
@@ -89,6 +86,13 @@ void fq_call_next(){
+/**
+ * Move to the next time in the function queue
+ */
+void fq_step() {
```
Would it be cleaner to just make fq_find_next() non-static and call it instead of adding this wrapper?
by the way, would you mind separating the move the fq_find_next() to another commit? (send both commits in this same PR, this would be just to keep things to their own)

aescolar commented on src/p2G4_pending_tx_list.c:
```diff
@@ -75,24 +75,3 @@ void txl_clear(uint d){
- */
-int txl_find_fitting_tx(p2G4_rx_t *rx_s, bs_time_t current_time){
-  register uint *used = tx_l_c.used;
-  for (int i = 0 ; i <= max_tx_nbr; i++) {
```
`max_tx_nbr` existed only for this. It can be removed now.
```diff
diff --git a/src/p2G4_pending_tx_list.c b/src/p2G4_pending_tx_list.c
index 4f8cf06..4fb2191 100755
--- a/src/p2G4_pending_tx_list.c
+++ b/src/p2G4_pending_tx_list.c
@@ -14,7 +14,6 @@ tx_l_c_t tx_l_c;
 static tx_el_t *tx_list = NULL;
 
 static uint nbr_devs;
-static int max_tx_nbr; //highest device transmitting at this point
 
 void txl_create(uint n_devs){
   tx_l_c.tx_list = bs_calloc(n_devs, sizeof(tx_el_t));
@@ -22,7 +21,6 @@ void txl_create(uint n_devs){
   tx_l_c.ctr = 0;
   tx_list = tx_l_c.tx_list;
   nbr_devs = n_devs;
-  max_tx_nbr = -1;
 }
 
 void txl_free(void){
@@ -53,7 +51,6 @@ void txl_register(uint d, p2G4_tx_t *tx_s, uint8_t* packet){
 void txl_activate(uint d){
   tx_l_c.used[d] = 1;
   tx_l_c.ctr++;
-  max_tx_nbr = BS_MAX(max_tx_nbr, ((int)d));
 }
 
 /**
@@ -66,12 +63,4 @@ void txl_clear(uint d){
     tx_list[d].packet = NULL;
   }
   tx_l_c.ctr++;
-
-  for (int i = max_tx_nbr; i >= 0 ; i--){
-    if (tx_l_c.used[i]){
-      max_tx_nbr = i;
-      return;
-    }
-  }
-  max_tx_nbr = -1;//if we didnt find any
 }
 ```

aescolar commented on src/p2G4_main.c:
```diff
@@ -137,8 +160,16 @@ static void rx_do_RSSI(uint d) {
   chm_RSSImeas(&tx_l_c, &RSSI_s, &rx_a[d].rx_done_s.rssi, d, current_time);
 }
 
+static inline void rx_queue_search_end(uint d){
```
silly thing: maybe better rx_*en*queue_search_end? (so it is clear it is a verb)

aescolar commented on src/p2G4_main.c:
```diff
@@ -161,6 +192,7 @@ static void f_rx_search(uint d) {
 
     bs_trace_raw_time(8,"RxDone (NoSync during Rx_search) for device %u\n", d);
 
+    rx_status->state = Rx_State_Inactive;
```
Not a functional problem, but just something which may be misleading: Many will end not going back to `Rx_State_Inactive` in many cases but be left in found.
May it be better to change `state` to be just searching/not_searching?

paulquevedojdrf replied:
I originally had that but for the corner case where two transmitters start at the same time the current behavior of lowest device number wins would be altered. If thats not an issue then a boolean would work.

aescolar replied:
Uhm, a change like this:
```diff
diff --git a/src/p2G4_main.c b/src/p2G4_main.c
index 5f9b3fa..2bf8e05 100644
--- a/src/p2G4_main.c
+++ b/src/p2G4_main.c
@@ -96,7 +96,7 @@ static void find_and_activate_rx(const p2G4_tx_t *tx_s, uint tx_d) {
            (rx_status->rx_s.radio_params.modulation & P2G4_MOD_SIMILAR_MASK))
     {
       rx_status->tx_nbr = tx_d;
-      rx_status->state = Rx_State_Found;
+      rx_status->state = Rx_State_NotSearching;
       fq_add(current_time, Rx_Found, rx_d);
     }
   }
@@ -192,7 +192,7 @@ static void f_rx_search(uint d) {
 
     bs_trace_raw_time(8,"RxDone (NoSync during Rx_search) for device %u\n", d);
 
-    rx_status->state = Rx_State_Inactive;
+    rx_status->state = Rx_State_NotSearching;
     rx_status->rx_done_s.end_time = current_time;
 
     p2G4_phy_resp_rx(d, &rx_status->rx_done_s);
diff --git a/src/p2G4_pending_tx_rx_list.h b/src/p2G4_pending_tx_rx_list.h
index 398b3c4..19e99a2 100755
--- a/src/p2G4_pending_tx_rx_list.h
+++ b/src/p2G4_pending_tx_rx_list.h
@@ -77,9 +77,8 @@ void txl_activate(uint dev_nbr);
  * Reception state
  */
 typedef enum {
-    Rx_State_Inactive = 0, //Not actively searching for any tx
+    Rx_State_NotSearching = 0, //Not actively searching for any tx
     Rx_State_Searching,//Searching for a fitting tx
-    Rx_State_Found,//Found a fitting tx and actively trying to receive it
 } rx_state_t;
```
Should not change the execution flow in any way (You still take it out of searching in `find_and_activate_rx()` )

aescolar commented on src/p2G4_main.c:
```diff
@@ -169,10 +201,15 @@ static void f_rx_search(uint d) {
 
-  if (tx_d >= 0) {
-    rx_status->tx_nbr = tx_d;
+  if (rx_status->tx_nbr >= 0) {
```
This if would be redundant now, or?

paulquevedojdrf replied:
yes copy paste leftover, although I suppose it doesn't hurt to avoid a possible out of bounds access

aescolar replied:
better remove it. It may confuse future readers, and that line should really never help

aescolar commented on src/p2G4_main.c:
```diff
@@ -82,6 +82,27 @@ static int pick_and_validate_abort(uint d, p2G4_abort_t *ab, const char* type) {
 
+/**
+ * Search all devices that are currently in the rx searching state.
+ * All devices able to receive this transmission begins the rx procedure
```
Maybe better change the comment like:
```diff
- * All devices able to receive this transmission begins the rx procedure
+ * All devices which may be able to receive this transmission are moved 
+ * to the Rx_Found state, where they may start synchronizing to it
```

aescolar commented on src/p2G4_main.c:
```diff
@@ -169,10 +201,15 @@ static void f_rx_search(uint d) {
-  if (tx_d >= 0) {
-    rx_status->tx_nbr = tx_d;
+  if (rx_status->tx_nbr >= 0) {
+    const uint tx_d = rx_status->tx_nbr;
```
Curiosity question: Did you see any difference in the resulting code from adding this const?

paulquevedojdrf replied:
nope, just a const correctness habit. I'll remove it

aescolar reviewed (changes requested):
Very nice.
Just a few minor things +
* Could you modify the commit title and add a description?
* We also need to update the svg in the doc

aescolar commented:
> Callgrinds showing same simulation profiled before and after this patch. Simulation results are identical.

Both profiles have dumping on :) (in the fast case it accounts for more than 12% of the runtime)
Anyhow, a 99.5% improvement in this use case with no drawbacks. Very nice indeed.

aescolar commented:
Do you feel like doing also the rx_payload change? (in a separate commit, it can be in this PR)
Something like this:
```diff --git a/src/p2G4_main.c b/src/p2G4_main.c
index 5f9b3fa..6f9dcfd 100644
--- a/src/p2G4_main.c
+++ b/src/p2G4_main.c
@@ -357,6 +357,9 @@ static void f_rx_payload(uint d){
     if (pick_and_validate_abort(d, &(rx_a[d].rx_s.abort), "Rx") != 0) {
       return;
     }
+    if ( rx_a[d].rx_s.abort.abort_time < rx_a[d].payload_end ) {
+        rx_a[d].payload_end = rx_a[d].rx_s.abort.abort_time;
+    }
   }
 
   if ( tx_l_c.used[rx_a[d].tx_nbr] == 0 ) { //if the Tx aborted
@@ -384,7 +387,12 @@ static void f_rx_payload(uint d){
     p2G4_handle_next_request(d);
     return;
   } else {
-    fq_add(current_time + 1, Rx_Payload, d);
+    bs_time_t next_time = current_time + 1;
+    if (rx_a[d].biterrors > 0) {
+      //If there is already bit errors, there is no need to keep recalculating, the result will be a CRC_ERROR
+      next_time = BS_MIN(rx_a[d].rx_s.abort.recheck_time, rx_a[d].payload_end);
+    }
+    fq_add(next_time, Rx_Payload, d);
     return;
   }
 }
```

aescolar committed:
doc: Update device interface svg to match new faster search

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
You can pick the updated svg from here: https://github.com/aescolar/ext_2G4_phy_v1/commits/with_doc

paulquevedojdrf commented:
> Do you feel like doing also the rx_payload change? (in a separate commit, it can be in this PR)
> Something like this:
> 
> ```diff
> index 5f9b3fa..6f9dcfd 100644
> --- a/src/p2G4_main.c
> +++ b/src/p2G4_main.c
> @@ -357,6 +357,9 @@ static void f_rx_payload(uint d){
>      if (pick_and_validate_abort(d, &(rx_a[d].rx_s.abort), "Rx") != 0) {
>        return;
>      }
> +    if ( rx_a[d].rx_s.abort.abort_time < rx_a[d].payload_end ) {
> +        rx_a[d].payload_end = rx_a[d].rx_s.abort.abort_time;
> +    }
>    }
>  
>    if ( tx_l_c.used[rx_a[d].tx_nbr] == 0 ) { //if the Tx aborted
> @@ -384,7 +387,12 @@ static void f_rx_payload(uint d){
>      p2G4_handle_next_request(d);
>      return;
>    } else {
> -    fq_add(current_time + 1, Rx_Payload, d);
> +    bs_time_t next_time = current_time + 1;
> +    if (rx_a[d].biterrors > 0) {
> +      //If there is already bit errors, there is no need to keep recalculating, the result will be a CRC_ERROR
> +      next_time = BS_MIN(rx_a[d].rx_s.abort.recheck_time, rx_a[d].payload_end);
> +    }
> +    fq_add(next_time, Rx_Payload, d);
>      return;
>    }
>  }
> ```

I looked at it but excluded it from this PR because that change yields different simulation results than the baseline so its a bit harder to  confirm its working correctly. As discussed, this is due to the change in call order to `random()`. 
I can do the change, but i think its better left as another PR.

aescolar commented:
>I looked at it but excluded it from this PR because that change yields different simulation results

Fair enough. Let's leave it for another PR. The gain is not so big anymore . Actually out of curiosity, after this PR, how much time is spent on the phy compared with the devices in your 150 devices use case?

paulquevedojdrf commented:
> Fair enough. Let's leave it for another PR. The gain is not so big anymore . Actually out of curiosity, after this PR, how much time is spent on the phy compared with the devices in your 150 devices use case?

top shows the phy using around 60% cpu, all the device instances hover around 1%

paulquevedojdrf committed:
optimization: call fq_find_next explicitly instead of on each enqueue/dequeue operation

Signed-off-by: Paul Quevedo <paulquevedo@jdrfelectromag.com>

aescolar commented on src/p2G4_main.c:
```diff
@@ -82,6 +82,27 @@ static int pick_and_validate_abort(uint d, p2G4_abort_t *ab, const char* type) {
+ * to the Rx_Found state, where they may start synchronizing to it
+ */
+static void find_and_activate_rx(const p2G4_tx_t *tx_s, uint tx_d) {
+  for (int rx_d = 0 ; rx_d < nbr_active_devs; rx_d++) {
```
There was a bug here, it should be `args.n_devs` not `nbr_active_devs`
(`nbr_active_devs` is just a count of how many devices are still there, but arrays/lists are not compacted when a device leaves -> we still need to iterate thru the whole array)

aescolar reviewed (commented):
I fixed that issue below, changed a bit the commit message and merged it.

aescolar commented:
The change is in now => closing the PR (please tell if you disagree about closing it)

I will not update the pointer in the Zephyr docker yet to give people a bit of time to try this change. Please tell if you find some issues with it.
